scripts/gen_attribute_completion_and_shape_inference_hpp.py

```diff
@@ -0,0 +1,322 @@
+
+
+def main():
+    template = """
+#ifndef MENOH_ATTRIBUTE_COMPLETION_AND_SHAPE_INFERENCE_HPP
+#define MENOH_ATTRIBUTE_COMPLETION_AND_SHAPE_INFERENCE_HPP
+/*
+ * This file is generated by {script_name}
+ * Do NOT modify this file directly
+ */
+#include <algorithm>
+#include <cassert>
+#include <string>
+#include <unordered_map>
+
+#include <menoh/array.hpp>
+#include <menoh/model_data.hpp>
+
+namespace menoh_impl {{
+    inline auto complete_attribute_and_infer_shape(
+            model_data& model_data,
+            std::unordered_map<std::string, array_profile> const&
+                input_profile_table) {{
+        using ints = std::vector<int>;
+        std::unordered_map<std::string, array_profile> variable_profile_table(
+            input_profile_table.begin(), input_profile_table.end());
+        std::transform(
+            model_data.parameter_name_and_array_list.begin(),
+            model_data.parameter_name_and_array_list.end(),
+            std::inserter(variable_profile_table,
+                          variable_profile_table.end()),
+            [](auto const& p){{
+                return std::make_pair(
+                    p.first,
+                    array_profile(p.second.dtype(), p.second.dims())); }});
+        auto profile_of = [&variable_profile_table](std::string const& name){{
+            assert(variable_profile_table.find(name) ==
+                variable_profile_table.end());
+            return variable_profile_table.at(name);
+        }};
+        auto dims_of = [&variable_profile_table, profile_of](
+            std::string const& name){{
+                return profile_of(name).dims();
+        }};
+        auto dtype_of = [&variable_profile_table, profile_of](
+            std::string const& name){{
+                return profile_of(name).dtype();
+        }};
+        auto ndims_of = [&dims_of](std::string const& parameter_name) {{
+            return dims_of(parameter_name).size();
+        }};
+        auto add_variable_to_table = [&variable_profile_table](
+            std::string const& name,
+            dtype_t dtype, ints const& dims){{
+                variable_profile_table.emplace(
+                    name, array_profile(dtype, dims));
+            }};
+
+        auto graph = make_graph(model_data.node_list); // FIXME reorder nodes
+        model_data.node_list = graph.node_list();
+        for(auto& node : model_data.node_list) {{
+            auto input = [&node](int i){{
+                return node.input_name_list.at(i);
+            }};
+            auto output = [&node](int i){{
+                return node.output_name_list.at(i);
+            }};
+            {code}
+            {unsupported_operator}
+        }}
+        return variable_profile_table;
+    }}
+}} // namespace menoh_impl
+
+#endif // MENOH_ATTRIBUTE_COMPLETION_AND_SHAPE_INFERENCE_HPP
+"""
+    code_list = []
+    code_list.append(make_completion_code("Abs"))
+    code_list.append(make_completion_code("Add"))
```

Comment:
Default `shape_inference_code` assumes an unary operator (it executes `assert(node.input_name_list.size() == 1)`), but `Add` is a binary operator.